beta 2 tutorial

Bradamant3 opened the pull request:
DO NOT MERGE. INCOMPLETE, ADDING PR FOR EARLY COMMENTS

hausdorff commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,163 @@
-Coming soon!
+# Tutorial
+
+The **ksonnet** [readme][readme] 
```
Might be worth considering making the value prop a little clearer up front to contextualize why this is important. Straw proposal to brainstorm:

```
ksonnet allows users to trivially add sidecars to Kubernetes applications (_e.g._, adding fluentd, linkerd, or istio) using _mixin libraries_. In this tutorial, we will:

* Introduce the concept of a mixin.
* Demonstrate how to write a mixin library that adds fluentd to a deployment object.
* Demonstrate how to use the mixin library in Kubernetes application code.
```

If we agree about the general structure of that intro, it would also seem to imply that we should structure the document differently, to concentrate less on building the deployment object up from scratch.

Thoughts?

Bradamant3 commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,163 @@
+(optional)
+* Created a test Kubernetes cluster
+
+NOTE: All import paths are relative to the root of the 
```
yes, will fix up

Bradamant3 replied:
Backed this one out, sort of. Added step to put tutorial files at root of ksonnet repo. Rationale: this way reader can copy bash literally without having to include path. Added note about -J flag and path if file is elsewhere.

Bradamant3 commented:
Yanked the container parts of the tutorial, rewrote tutorial outline, added some minimal explanation of mixins to readme. Please note all items marked with TODO (not all in comments but using that convention so I can find them easily). LMK what else you need from me.

Bradamant3 commented:
Another approach -- longer examples for built-in mixins, based on a trimmed-down version of the ingress controller in Chuck's work on the Heptio cluster config

Bradamant3 commented:
Deployment examples are now complete with bash scripts to call and run, YAML output. Waiting now on fluentd mixin from Alex.

hausdorff commented on README.md:
```diff
@@ -162,14 +163,24 @@ abstraction they approach. For most users, the right entry point is:
-spec, and are organized by the API groups they belong to:
+that extends the object model and functions of **Jsonnet** to implement
+the Kubernetes API. These are generated directly from the Swagger
+definition, and are organized by the API groups they belong to:
```
I believe this should probably be "OpenAPI specification"

Bradamant3 replied:
yup, still to fix.

hausdorff commented on README.md:
```diff
@@ -162,14 +163,24 @@ abstraction they approach. For most users, the right entry point is:
 
+## Mixins
+
+Mixins are a core feature of **Ksonnet**. Conceptually, they provide dynamic inheritance, at 
```
Just FYI, I believe that marketing have decided that we should lowercase the first letter of ksonnet.

hausdorff commented on README.md:
```diff
@@ -162,14 +163,24 @@ abstraction they approach. For most users, the right entry point is:
+create new ones.
+
+**Ksonnet** ships with a large library of built-in mixins, or you can write your own custom mixins. 
+The [tutorial](/docs/TUTORIAL.md) shows you how to create a custom mixin that you can then 
```
I bet you've already thought of this, so just for my own education: this is meant to work on the website, right?

abiogenesis-now commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,305 @@
+
+**ksonnet** lets you configure or modify any Kubernetes object. In the following 
+exercises, you create an explicit IngressController and Service from the nginx 
+deployment described in the [readme]. You simplify the 
```
to avoid mentioning the readme multiple times, maybe you can mention the nginx deployment in the prereqs:

e.g. 
> * Created a test Kubernetes cluster with a nginx deployment

and then change this sentence to:

> By using **ksonnet**'s built-in mixins, you can simplify the process of  defining and adding new objects. In the following exercises, for example, you use mixins to create an explicit IngressController and Service from the original nginx deployment.

abiogenesis-now commented on docs/TUTORIAL.md:
````diff
@@ -1,3 +1,305 @@
+
+First, define base configuration values, dependent objects, and the IngressController:
+
+```javascript
````
I've sometimes seen people link to Github gists---do we want to do this, or assume that people will simply copy + paste? I'm ok either way.

abiogenesis-now commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,305 @@
+local deployment = k.extensions.v1beta1.deployment;
+local container = deployment.mixin.spec.template.spec.containersType;
+
+// base configuration. These values can be passed explicitly instead, but this 
```
I think I understand what you mean, but can you maybe preface the second sentence with "In the code below,"?

abiogenesis-now commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,305 @@
+    deployment.mixin.spec.template.metadata.annotations(conf.ingressController.annotations) +
+    deployment.mixin.spec.template.spec.terminationGracePeriodSeconds(conf.ingressController.terminationGracePeriodSeconds);
+
+// call the deployment
```
hmm...could we say "return the deployment object" instead? "calling" makes me feel like we're running it, but it's really just going to return and render

abiogenesis-now commented on docs/TUTORIAL.md:
````diff
@@ -1,3 +1,305 @@
+
+```bash
+jsonnet ingressController.jsonnet > deployment.yaml
+kubectl apply -f deployment.yaml
````
can you move the `kubectl` command to *after* the part where you show the generated YAML? Since deploying to the cluster itself feels like a different step.

abiogenesis-now commented on docs/TUTORIAL.md:
````diff
@@ -1,3 +1,305 @@
+kind: List
+```
+
+Now you can add other objects simply, using additional built-in mixins. The Service 
````
This might be just personal preference, so you can take with a grain of salt, but consider having "Ingress controller" and "Service definition" subheaders to denote the separate setups?

abiogenesis-now commented on docs/TUTORIAL.md:
````diff
@@ -1,3 +1,305 @@
+definition looks like this:
+
+```javascript
+local ingressService = service.new(conf.ingressController.name, conf.ingressController.selector, [
````
I wonder if we can make the indentations a bit nicer here

abiogenesis-now commented on docs/TUTORIAL.md:
````diff
@@ -1,3 +1,305 @@
+    service.mixin.spec.type("LoadBalancer");
+```
+
+After you add the Service to your configuration, the completed file looks like this:
````
s/the completed file/your modified `ingressController.jsonnet` ?

abiogenesis-now replied:
I wish we could highlight the new lines in a different color but github so ugh

abiogenesis-now commented on docs/TUTORIAL.md:
````diff
@@ -1,3 +1,305 @@
+
+```bash
+jsonnet ingressController.jsonnet > deployment.yaml
+kubectl apply -f deployment.yaml
````
same comment as above

abiogenesis-now commented on docs/TUTORIAL.md:
````diff
@@ -1,3 +1,305 @@
+kind: List
+```
+
+You can also easily add the built-in mixins for liveness and readiness probes to the container object.
````
do you have links for these?

abiogenesis-now commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,305 @@
+* Write a custom mixin library for Fluentd 
+* Add the custom Fluentd mixin to your application deployemtn
+
+## Prerequisites
```
I like numbering the steps (so they seem sequential, rather than potentially unrelated sections) but I think I remember you having the opposite opinion ;P

abiogenesis-now commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,305 @@
+to an existing Kubernetes cluster. This tutorial shows you how to:
+
+* Modify the deployment using the built-in **ksonnet** mixins
+* Write a custom mixin library for Fluentd 
```
can you link Fluentd and add a brief description (e.g. short phrase) of what it is? since we don't want to assume that everyone has the background

abiogenesis-now commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,305 @@
+Because mixins let you assemble components at runtime (dynamic inheritance),
+this approach means that different team members can define the 
+Kubernetes objects that your cluster needs. You can then mix them into 
+your master configuration without having to copy all the details.
```
maybe be more specific:
> You can mix them into your master configuration using imports, without having to copy all of the details.

abiogenesis-now commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,305 @@
+
+You can also easily add the built-in mixins for liveness and readiness probes to the container object.
+
+## Write a custom mixin to include as a sidecar
```
maybe just title this "Write a custom fluentd mixin"?

hausdorff replied:
Yeah, it's not really a sidecar if it's in a `DaemonSet`.

Bradamant3 commented:
Jessica's comments not yet addressed. The mixin is broken down into its top-level component parts, although I start with the DaemonSet alone because it's sort of the core object. Then we move to parts, then util (to clear up a few things, then I'm not sure whether app or mixin yet, but you get the idea.

hausdorff commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,413 @@
+Your application needs logging and monitoring, so you can write a mixin to define a 
+Fluentd container. You can then add it to your cluster as a sidecar.
+
+Because mixins let you assemble components at runtime (dynamic inheritance),
```
"Dynamic inheritance" might be a little confusing to people who are not in the know. It might be better to talk about it like a fancy object merge.

hausdorff commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,413 @@
+other parts that we'll use to assemble higher-level definitions.
+
+Note in particular that the base DaemonSet definition doesn't include RBAC support, which we 
+need, so we add it to the collection of parts:
```
I like the structure, I think it will be more effective if we start by showing how to use the parts, rather than their definitions.

Bradamant3 commented:
Still to fix up, example deployment and service in first section.

Bradamant3 commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,305 @@
 # Tutorial
 
-Coming soon!
+The **ksonnet** [readme][readme] 
```
But this assumes we know that the reader worked through the readme example first. We can't know. (Another reason for simple present)

abiogenesis-now replied:
hmm...I think of it as indicative as a reader (e.g. oh they think I should have read that? I haven't so maybe I should open it up in a tab just in case). However, this is personal preference so if you feel strongly I'm fine leaving as is.

hausdorff commented on docs/TUTORIAL.md:
```diff
@@ -1,3 +1,310 @@
+NOTE: All import paths are relative to the root of the 
+**ksonnet** repository.
+
+## Add a service to the deployment configuration
```
For new the big thing in this draft is: it sends like if we use normal json for the service and the deployment, then we can skip right to explaining how the missions work.
